client/allocrunner/tasklifecycle/doc.go

```diff
@@ -0,0 +1,90 @@
+blocking the main tasks.
+
+                     ┌────────┐
+                     │ ALLOC  │
+                     │ RUNNER │
+                     └───┬────┘
+                         │
+                     Task state
+                         │
+            ┌────────────▼────────────┐
+            │Current state:           │
+            │Prestart                 │         ┌─────────────┐
+            │                         │         │ TASK RUNNER │
+            │     ┌───────────────────┼─────────┤ (Prestart)  │
+            │     │                   │         └─────────────┘
+            │     │                   │
+            │     │                   │         ┌─────────────┐
+            │     │ COORDINATOR       │         │ TASK RUNNER │
+            │     │             ┌─ ─ ─┼─ ─ ─ ─┬╶┤   (Main)    │
+            │     │             ╷     │       ╷ └─────────────┘
+            │     │             ╷     │       ╷
+            │     │             ╷     │       ╷ ┌─────────────┐
+            │   Prestart       Main   │       ╷ │ TASK RUNNER │
+            └─────┬─┬───────────┬─┬───┘       └╶┤   (Main)    │
+                  │ │Open/      ╷ │Open/        └─────────────┘
+                  │ │Close      ╷ │Close
+               ┌──┴─▼─┐      ┌──┴─▼─┐
+               │ GATE │      │ GATE │
+               └──────┘      └──────┘
+
+
+When the prestart task completes, the allocRunner will send a new batch of task
+states to the Coordinator that will cause it to transition to a state where it
+will close the Gate for prestart tasks, blocking their execution, and will open
+the Gate for main tasks, allowing them to start.
+
+                     ┌────────┐
+                     │ ALLOC  │
+                     │ RUNNER │
+                     └───┬────┘
+                         │
+                     Task state
+                         │
+            ┌────────────▼────────────┐
+            │Current state:           │
+            │Main                     │         ┌─────────────┐
+            │                         │         │ TASK RUNNER │
+            │     ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┼─ ─ ─ ─ ─┤ (Prestart)  │
+            │     ╷                   │         └─────────────┘
+            │     ╷                   │
+            │     ╷                   │         ┌─────────────┐
+            │     ╷ COORDINATOR       │         │ TASK RUNNER │
+            │     ╷             ┌─────┼───────┬─┤   (Main)    │
+            │     ╷             │     │       │ └─────────────┘
+            │     ╷             │     │       │
+            │     ╷             │     │       │ ┌─────────────┐
+            │   Prestart       Main   │       │ │ TASK RUNNER │
+            └─────┼─┬───────────┬─┬───┘       └─┤   (Main)    │
+                  ╷ │Open/      │ │Open/        └─────────────┘
+                  ╷ │Close      │ │Close
+               ┌──┴─▼─┐      ┌──┴─▼─┐
+               │ GATE │      │ GATE │
+               └──────┘      └──────┘
+*/
```

Comment:
I used [ASCIIFlow](https://asciiflow.com/) since it was the first result when I searched for ASCII diagram tool 😬 

I had to edit the result a bit to make the dashed like more distinguishable, but I will link to what I have.